sound/soc/sof/loader.c

```diff
@@ -129,6 +130,85 @@ int snd_sof_fw_parse_ext_data(struct snd_sof_dev *sdev, u32 bar, u32 offset)
 }
 EXPORT_SYMBOL(snd_sof_fw_parse_ext_data);
 
+static size_t snd_sof_ext_man_size(const struct firmware *fw)
+{
+	const struct ext_man_header *head = (void *)fw->data;
+
+	if (fw->size < sizeof(*head))
+		return 0;
+
+	/*
+	 * When fw points to extended manifest,
+	 * then first u32 must be equal EXT_MAN_MAGIC_NUMBER.
+	 */
+	if (head->magic == EXT_MAN_MAGIC_NUMBER)
+		return head->full_size;
+
+	return 0;
+}
+
+/* parse extended FW manifest data structures */
+static int snd_sof_fw_ext_man_parse(struct snd_sof_dev *sdev,
+				    const struct firmware *fw)
+{
+	const struct ext_man_header *head = (struct ext_man_header *)fw->data;
+	const struct ext_man_elem_header *elem_hdr;
+	ssize_t remaining;
+	uintptr_t iptr;
+	int ret = 0;
+
+	remaining = head->full_size - head->header_size;
+
+	/* Assert firmware starts with extended manifest */
+	if (snd_sof_ext_man_size(fw) == 0)
+		return -EINVAL;
+
+	/* in compatible version */
+	if (EXT_MAN_VERSION_INCOMPATIBLE(EXT_MAN_VERSION, head->header_version)) {
+		dev_err(sdev->dev, "error: extended manifest version 0x%X differ from used 0x%X\n",
+			head->header_version, EXT_MAN_VERSION);
+		return -EINVAL;
+	}
+
+	/* get first header */
+	iptr = (uintptr_t)fw->data + head->header_size;
```

Comment:
@ktrzcinx this comment is a bit misleading, maybe "get first element header" ... I'd argue the first header is at zero and that's the "ext_man_header"

Reply:
changed